src/dos/dos.cpp

```diff
@@ -504,13 +504,24 @@ static Bitu DOS_21Handler(void) {
 		break;
 	}
 	case 0x2d:		/* Set System Time */
-		LOG(LOG_DOSMISC,LOG_ERROR)("DOS:Set System Time not supported");
-		//Check input parameters nonetheless
 		if( reg_ch > 23 || reg_cl > 59 || reg_dh > 59 || reg_dl > 99 )
 			reg_al = 0xff; 
 		else { //Allow time to be set to zero. Restore the orginal time for all other parameters. (QuickBasic)
 			if (reg_cx == 0 && reg_dx == 0) {time_start = mem_readd(BIOS_TIMER);LOG_MSG("Warning: game messes with DOS time!");}
 			else time_start = 0;
+			// Original IBM PC used ~1.19MHz crystal for timer,
+			// because at 1.19MHz, 2^16 ticks is ~1 hour, making it
+			// easy to count hours and days. More precisely:
+			//
+			// clock updates at 1193180/65536 ticks per second.
+			// ticks per second ≈ 18.2
+			// ticks per hour   ≈ 65543
+			// ticks per day    ≈ 1573040
+			constexpr uint64_t ticks_per_day = 1573040;
+			uint32_t ticks = (uint32_t)(
+			        ((double)(reg_ch * 3600 + reg_cl * 60 + reg_dh)) *
+			        ticks_per_day / (24 * 3600));
```

Comment:
I think we can keep it integer to too:
```c++
		constexpr int64_t ticks_per_day = 1573040;
		const auto seconds = reg_ch * 3600 + reg_cl * 60 + reg_dh;
		const auto ticks = ticks_per_day * seconds / (24 * 3600);
		mem_writed(BIOS_TIMER, check_cast<uint32_t>(ticks));
```

That `check_cast`<uint32_t>(ticks)` is used when we step down to a narrower type; in this case asserting that the (signed 64-bit) `ticks` falls between 0 and UINT32_MAX. It's just an assert, so is only active in non-release builds.

Reply:
Wrote a quick test-wrapper that compares the original against the int-only:

https://onlinegdb.com/GWVsTiQU6h

```text
00:00:00 to ticks = 0 | 0
00:00:30 to ticks = 546 | 546
00:10:00 to ticks = 10923 | 10923
00:10:30 to ticks = 11470 | 11470
00:20:00 to ticks = 21847 | 21847
00:20:30 to ticks = 22393 | 22393
00:30:00 to ticks = 32771 | 32771
00:30:30 to ticks = 33317 | 33317
00:40:00 to ticks = 43695 | 43695
00:40:30 to ticks = 44241 | 44241
00:50:00 to ticks = 54619 | 54619
00:50:30 to ticks = 55165 | 55165
04:00:00 to ticks = 262173 | 262173
04:00:30 to ticks = 262719 | 262719
04:10:00 to ticks = 273097 | 273097
04:10:30 to ticks = 273643 | 273643
04:20:00 to ticks = 284021 | 284021
04:20:30 to ticks = 284567 | 284567
04:30:00 to ticks = 294945 | 294945
04:30:30 to ticks = 295491 | 295491
04:40:00 to ticks = 305868 | 305868
04:40:30 to ticks = 306415 | 306415
04:50:00 to ticks = 316792 | 316792
04:50:30 to ticks = 317338 | 317338
08:00:00 to ticks = 524346 | 524346
08:00:30 to ticks = 524892 | 524892
08:10:00 to ticks = 535270 | 535270
08:10:30 to ticks = 535816 | 535816
08:20:00 to ticks = 546194 | 546194
08:20:30 to ticks = 546740 | 546740
08:30:00 to ticks = 557118 | 557118
08:30:30 to ticks = 557664 | 557664
08:40:00 to ticks = 568042 | 568042
08:40:30 to ticks = 568588 | 568588
08:50:00 to ticks = 578966 | 578966
08:50:30 to ticks = 579512 | 579512
12:00:00 to ticks = 786520 | 786520
12:00:30 to ticks = 787066 | 787066
12:10:00 to ticks = 797443 | 797443
12:10:30 to ticks = 797990 | 797990
12:20:00 to ticks = 808367 | 808367
12:20:30 to ticks = 808913 | 808913
12:30:00 to ticks = 819291 | 819291
12:30:30 to ticks = 819837 | 819837
12:40:00 to ticks = 830215 | 830215
12:40:30 to ticks = 830761 | 830761
12:50:00 to ticks = 841139 | 841139
12:50:30 to ticks = 841685 | 841685
16:00:00 to ticks = 1048693 | 1048693
16:00:30 to ticks = 1049239 | 1049239
16:10:00 to ticks = 1059617 | 1059617
16:10:30 to ticks = 1060163 | 1060163
16:20:00 to ticks = 1070541 | 1070541
16:20:30 to ticks = 1071087 | 1071087
16:30:00 to ticks = 1081465 | 1081465
16:30:30 to ticks = 1082011 | 1082011
16:40:00 to ticks = 1092388 | 1092388
16:40:30 to ticks = 1092935 | 1092935
16:50:00 to ticks = 1103312 | 1103312
16:50:30 to ticks = 1103858 | 1103858
20:00:00 to ticks = 1310866 | 1310866
20:00:30 to ticks = 1311412 | 1311412
20:10:00 to ticks = 1321790 | 1321790
20:10:30 to ticks = 1322336 | 1322336
20:20:00 to ticks = 1332714 | 1332714
20:20:30 to ticks = 1333260 | 1333260
20:30:00 to ticks = 1343638 | 1343638
20:30:30 to ticks = 1344184 | 1344184
20:40:00 to ticks = 1354562 | 1354562
20:40:30 to ticks = 1355108 | 1355108
20:50:00 to ticks = 1365486 | 1365486
20:50:30 to ticks = 1366032 | 1366032
```

Reply:
@kcgen I have already changed it to check_cast code as you mentioned. Thanks.